datafusion/src/physical_plan/window_frames.rs

```diff
@@ -0,0 +1,200 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+//! Window frame
+//!
+//! The frame-spec determines which output rows are read by an aggregate window function. The frame-spec consists of four parts:
+//! - A frame type - either ROWS, RANGE or GROUPS,
+//! - A starting frame boundary,
+//! - An ending frame boundary,
+//! - An EXCLUDE clause.
+
+use crate::error::{DataFusionError, Result};
+use sqlparser::ast::{WindowFrame, WindowFrameBound, WindowFrameUnits};
+
+const DEFAULT_WINDOW_FRAME: WindowFrame = WindowFrame {
+    units: WindowFrameUnits::Range,
+    start_bound: WindowFrameBound::Preceding(None),
+    end_bound: Some(WindowFrameBound::CurrentRow),
+};
+
+fn get_bound_rank(bound: &WindowFrameBound) -> (u8, u64) {
+    match bound {
+        WindowFrameBound::Preceding(None) => (0, 0),
+        WindowFrameBound::Following(None) => (4, 0),
+        WindowFrameBound::Preceding(Some(0))
+        | WindowFrameBound::CurrentRow
+        | WindowFrameBound::Following(Some(0)) => (2, 0),
+        WindowFrameBound::Preceding(Some(v)) => (1, u64::MAX - *v),
+        WindowFrameBound::Following(Some(v)) => (3, *v),
+    }
+}
+
+/// Validate a window frame if present, otherwise return the default window frame.
+pub(crate) fn validate_window_frame(
```

Comment:
In general, it would be cool if this validation was done as part of the SQL AST --> LogicalPlan / Expr. Then the rest of the optimizers could assume that the structures were valid. 

if we added a DataFusion specific `WindowFrame` struct and then performed this validation while constructing it, by the time the physical plan saw a Window Frame it would be in a known good state

Reply:
see https://github.com/apache/arrow-datafusion/pull/506